sources/api/thar-be-updates/src/main.rs

```diff
@@ -0,0 +1,427 @@
+fn prepare_update(status: &mut UpdateStatus<update_metadata::Update>) -> Result<()> {
+    debug!("Spawning 'updog update-image'");
+    let chosen_update = status.chosen_update().context(error::ChosenUpdate)?.clone();
+    let output = Command::new("updog")
+        .arg("update-image")
+        .output()
+        .context(error::Updog)?;
+    status.set_recent_command_info(UpdateCommand::PrepareUpdate, &output);
+    if output.status.success() {
+        status.set_staging_partition_image_info(chosen_update);
+    } else {
+        warn!("Failed to prepare the update with updog");
+    }
+    Ok(())
+}
+
+/// "Activates" the staged update by letting updog set up the appropriate boot flags
+fn activate_update(status: &mut UpdateStatus<update_metadata::Update>) -> Result<()> {
+    debug!("Spawning 'updog update-apply'");
+    let output = Command::new("updog")
+        .arg("update-apply")
+        .output()
+        .context(error::Updog)?;
+    status.set_recent_command_info(UpdateCommand::ActivateUpdate, &output);
+    if output.status.success() {
+        status.mark_staging_partition_next_to_boot()?;
+    } else {
+        warn!("Failed to activate the update with updog");
+    }
+    Ok(())
+}
+
+/// "Deactivates" the staged update by rolling back actions done by `activate_update`
+fn deactivate_update(status: &mut UpdateStatus<update_metadata::Update>) -> Result<()> {
+    debug!("Spawning 'updog revert-update-apply'");
+    let output = Command::new("updog")
+        .arg("revert-update-apply")
+        .output()
+        .context(error::Updog)?;
+    status.set_recent_command_info(UpdateCommand::DeactivateUpdate, &output);
+    if output.status.success() {
+        status.unmark_staging_partition_next_to_boot()?;
+    } else {
+        warn!("Failed to deactivate the update with updog");
+    }
+    Ok(())
+}
+
+fn run() -> Result<()> {
+    // Parse and store the args passed to the program
+    let args = parse_args(env::args());
+
+    // TerminalMode::Mixed will send errors to stderr and anything less to stdout.
+    TermLogger::init(args.log_level, LogConfig::default(), TerminalMode::Mixed)
+        .context(error::Logger)?;
+
+    // Open the lockfile for concurrency control, create it if it doesn't exist
+    let lockfile = File::create(UPDATE_LOCKFILE).context(error::LockFile {
+        path: UPDATE_LOCKFILE,
+    })?;
+
+    // Check if the update status file exists. If it doesn't, create and initialize it.
+    if !Path::new(UPDATE_STATUS_FILE).is_file() {
+        // Get an exclusive lock for creating the update status file
+        lock_exclusive(&lockfile)?;
+        initialize_update_status()?;
+        unlock(&lockfile)?;
+    }
+
+    match args.subcommand {
+        // 'refresh-updates' is allowed under every update state
+        UpdateCommand::RefreshUpdates => {
+            lock_exclusive(&lockfile)?;
+            let mut update_status = get_update_status()?;
+            match fork() {
```

Comment:
`apiserver` needs `t-b-u` to return quickly because it cannot afford to block and wait on a result when it dispatches an update command. Hence the `fork` here to create a child process to handle spawning the `updog` process and updating the status once that's done.

Reply:
I see, it's kind of mind-bending but cool.